srml/contract/src/lib.rs

```diff
@@ -195,16 +202,25 @@ pub struct RawAliveContractInfo<CodeHash, Balance, BlockNumber> {
 	pub deduct_block: BlockNumber,
 }
 
-#[derive(Encode, Decode)]
-pub struct TombstoneContractInfo<T: Trait>(T::Hash);
+pub type TombstoneContractInfo<T> =
+	RawTombstoneContractInfo<<T as system::Trait>::Hash, <T as system::Trait>::Hashing>;
+
+// Workaround for https://github.com/rust-lang/rust/issues/26925 . Remove when sorted.
+#[derive(Encode, Decode, PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct RawTombstoneContractInfo<H, Hasher>(H, PhantomData<Hasher>);
 
-impl<T: Trait> TombstoneContractInfo<T> {
-	fn new(storage_root: Vec<u8>, storage_size: u64, code_hash: CodeHash<T>) -> Self {
+impl<H, Hasher> RawTombstoneContractInfo<H, Hasher>
+where
+	H: Member + MaybeSerializeDebug + AsRef<[u8]> + AsMut<[u8]> + Copy + Default + rstd::hash::Hash,
+	Hasher: Hash<Output=H>,
+{
+	fn new(storage_root: Vec<u8>, storage_size: u64, code_hash: H) -> Self {
 		let mut buf = Vec::new();
 		storage_root.using_encoded(|encoded| buf.extend_from_slice(encoded));
 		storage_size.using_encoded(|encoded| buf.extend_from_slice(encoded));
```

Comment:
I wonder if it is necessary to put storage size in the tombstone, having a storage root identical with different storage size does not seems possible to me (or at least should not be possible). I do not really see the value of putting storage_size in tombstone, except that it can be a burden for runtime update (in fact storage size way could not be fix at all as we cannot really restore tombstone (if the storage size was stored unhashed in the tombstone it would be possible)).

Reply:
Yes I agree, 

> it can be a burden for runtime update

hmm interesting.

Also that makes me think the storage_size_offset is added in storage size at creation of the contract. maybe it would be prefered to add it when computing rent actually.